dev/com.ibm.ws.security.wim.adapter.ldap_fat.krb5/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/krb5/ApacheDSandKDC.java

```diff
@@ -0,0 +1,649 @@
+                 */
+                TgtRequest clientTgtReq = new TgtRequest();
+                clientTgtReq.setClientPrincipal(bindPrincipalName);
+                clientTgtReq.setPassword(bindPassword);
+                clientTgtReq.setRenewable(true);
+
+                TgTicket tgt = conn.getTgt(clientTgtReq);
+
+                credCache = new CredentialsCache();
+                PrincipalName princ = new PrincipalName(bindPrincipalName, PrincipalNameType.KRB_NT_PRINCIPAL);
+                princ.setRealm(tgt.getRealm());
+                credCache.setPrimaryPrincipalName(princ);
+                Credentials cred = new Credentials(tgt);
+                credCache.addCredentials(cred);
+                CredentialsCache.store(ccFile, credCache);
+            } else {
+                Kinit kinit = new Kinit(conn);
+                kinit.setCredCacheFile(ccFile);
+
+                kinit.kinit(bindPrincipalName, bindPassword);
+
+                credCache = CredentialsCache.load(ccFile);
+            }
+
+            assertNotNull("TicketCache failed to create", credCache);
+
+        } finally {
+            kdcConfig.setMaximumTicketLifetime(prevMaxTicket);
+            kdcConfig.setRenewableAllowed(previousRenew);
+            kdcConfig.setMinimumTicketLifetime(prevMin);
+        }
+
+        Log.info(c, "createTicketCacheShortLife", "Created ticket cache: " + ticketCacheFile);
+
+        return ccFile.getAbsolutePath();
+
+    }
+
+    /**
+     * Create a keytab file with the default bindUser and password
+     *
+     * @throws Exception
+     */
+    public static void createKeyTabFile() throws Exception {
+        Log.info(c, "createKeyTabFile", "Creating keytab for " + bindPrincipalName);
+        File keyTabTemp = File.createTempFile(bindUserName + "-", ".keytab");
+        if (!FAT_TEST_LOCALRUN) {
+            keyTabTemp.deleteOnExit();
+        }
+
+        KerberosTime timeStamp = new KerberosTime();
+        int principalType = 1; // KRB5_NT_PRINCIPAL
+
+        Keytab keytab = Keytab.getInstance();
+
+        List<KeytabEntry> entries = new ArrayList<KeytabEntry>();
+
+        for (Map.Entry<EncryptionType, EncryptionKey> keyEntry : KerberosKeyFactory.getKerberosKeys(
+                                                                                                    bindPrincipalName, bindPassword)
+                        .entrySet()) {
+            final EncryptionKey key = keyEntry.getValue();
+            final byte keyVersion = (byte) key.getKeyVersion();
+            entries.add(new KeytabEntry(bindPrincipalName, principalType, timeStamp, keyVersion, key));
+        }
+
+        keytab.setEntries(entries);
+        keytab.write(keyTabTemp);
+
+        keytabFile = keyTabTemp.getAbsolutePath();
+
+        Log.info(c, "createKeyTabFile", "Created keytab: " + keytabFile);
+        Log.info(c, "createKeyTabFile", "Keytab contents: " + FileUtils.readFile(keytabFile));
```

Comment:
Is this human readable? May just need to print the keytabentries?

Reply:
Not really -- will print out the entries.